sprokit/processes/core/shift_detected_object_set_frames_process.cxx

```diff
@@ -0,0 +1,163 @@
+ *
+ *  * Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ *  * Neither name of Kitware, Inc. nor the names of any contributors may be used
+ *    to endorse or promote products derived from this software without specific
+ *    prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "shift_detected_object_set_frames_process.h"
+
+#include <vital/config/config_block.h>
+#include <sprokit/processes/kwiver_type_traits.h>
+#include <sprokit/pipeline/process_exception.h>
+
+/**
+ * \file shift_detected_object_set_frames_process.cxx
+ *
+ * \brief Implementation of the detected object set frame shift process
+ */
+
+namespace kwiver
+{
+
+class shift_detected_object_set_frames_process::priv
+{
+  public:
+    typedef int32_t number_t;
+    priv(number_t offset);
+    ~priv();
+
+    number_t remaining_offset;
+
+    static vital::config_block_key_t const config_value;
+    static vital::config_block_value_t const default_value;
+
+    static vital::detected_object_set_sptr const empty_detected_object_set_sptr;
+};
+
+vital::config_block_key_t const
+shift_detected_object_set_frames_process::priv::config_value =\
+  vital::config_block_key_t("offset");
+vital::config_block_value_t const
+shift_detected_object_set_frames_process::priv::default_value =\
+  vital::config_block_value_t("0");
+
+vital::detected_object_set_sptr const
+shift_detected_object_set_frames_process::priv::empty_detected_object_set_sptr =\
+  std::make_shared<vital::detected_object_set>();
+
+
+shift_detected_object_set_frames_process
+::shift_detected_object_set_frames_process(vital::config_block_sptr const& config)
+  : process(config)
+  , d()
+{
+  make_ports();
+  make_config();
+}
+
+shift_detected_object_set_frames_process
+::~shift_detected_object_set_frames_process()
+{
+}
+
+void
+shift_detected_object_set_frames_process
+::make_config()
+{
+  declare_configuration_key(
+    priv::config_value,
+    priv::default_value,
+    vital::config_block_description_t("The offset to shift the input by."));
```

Comment:
Ah, that's certainly cleaner.  Are there reasons to prefer one approach to the other apart from that?

Reply:
Mostly because it is cleaner and less code, and less code means fewer errors. The original sprokit pattern for processes required many source lines and also required the data type to be entered more than once. Using the trait approach centralizes all the config item details. Also the traits can be shared among other processes to give some consistency for config items. The original port pattern from sprokit was equally as verbose. One of the major goals with the port traits was to have them shared between all processes to centralize and standardize port names and types.

Reply:
Gotcha.  Thanks for the explanation, that makes sense.